keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,432 @@
+
+### Goals
+
+This proposal aims to:
+
+* Allow developers to define resource limits on the pod level in addition to the individual container level.
+* Use as much as possible the Linux kernel's cgroup ability to define limits in hierarchies to enforce the defined limits.
+
+### Non-Goals
+
+*  Providing a general-purpose interface to the full range of possible resource management provided by the Linux cgroup hierarchy.
+
+## Proposal
+
+The Pod QoS enhancement already implemented in Kubernetes manages resources as a hierarchy of cgroups in the following way:
+
+<pre>
+  kubepods
+   |
+   +-- Guaranteed-pod0
+   |   |
+   |   +-- container0 (pause container)
+   |   |
+   |   +-- container1 (first container)
+   |   |
+   |  = ... =
+   |   |
+   |   +-- containerN (N-th container)
+   |
+   +-- QoS CGroup (one of burstable, or besteffort)
+       |
+       \ pod 
+          |
+          +-- container0 (pause container)
+          |
+          +-- container1 (first container)
+          |
+        = ... =
+          |
+          +-- containerN (N-th container)
+</pre>
+
+The proposal in this KEP is to allow users to define resources on the Pod level itself. Both resource requests and limits are allowed. The scheduler
+algorithm will consider the pod-level resources if they exist in preference to the container-level resources. If no pod-level resource section is provided,
+then the scheduler will continue to operate as it does in the current implementation. The QoS definitions will be updated to consider the pod-level resources
+when classifying a pod into one of the QoS levels. If the pod level doesn't include any resource limits, then kubelet will continue to function as it does today.
+
+The current implementation will set the cgroup limits for the memory/CPU resources on the pod level of the hierarchy only in the following cases:
+1. The QoS level is `Guaranteed`. 
+1. The QoS level is `Burstable` and all of the containers specify a limit for the relevant resource. 
+
+In the current implementation pods are considered `Guaranteed` if and only if all of the included containers define a resource request which is equal
+to the resource limit. The current implementation considers a pod to be `Burstable` if at least one container defines resource requests or limits. Any
+pods not matching the `Guaranteed` or `Burstable` QoS definitions are placed in the `BestEffort` QoS level.
+
+This proposal suggests modifying these definitions the following: 
+
+A pod is considered `Guaranteed` if either of the following two conditions apply: 
+1. All of the containers included in the pod define memory and CPU resource requests equal to resource limits
+1. The pod includes a pod-level memory and CPU resource request which is equal to the resource limit, even if one or more of the included containers doesn't define either a CPU or a memory request or limit.
```

Comment:
>  * must the pod value be >= sum(containers[*].value)? Or is it valid to specify ONLY pod resources and no container resources? What if my containers each ask for 100 CPUs and my pod asks for 1 CPU? Is that valid (proportional?) or an error?

It is completely valid to specify only pod level resources. All of the containers will be restrained by the Linux cgroups relative to the pod level cgroup.

Each Linux cgroup behaves differently in cases where the lower-level requests exceed a higher point in the hierarchy. I'll update the KEP with the relevant information for the CPU and memory cgroups.

>  *    do we plan to reify the pod value or just infer it? E.g. we could add it to defaulting - sum up to containers and fill in the pod values

Defaulting the pod-level resources would be very wrong as this would potentially change the way that CPU quota is distributed to the containers. The Linux CPU cgroup always treats each level in the hierarchy relative to its parent. Imposing an additional limit when it was not explicitly requested can have unintended consequences. Limits should always be explicitly defined by the developer, not defaulted by the kubelet. Depending on the specific cgroup, not setting a limit might be different than setting a specific limit at a specific level in the cgroup hierarchy.

Reply:
> It is completely valid to specify only pod level resources

We should document this really clearly.    We frequently say or imply that not setting a limit  means "no limit".  What it really means is "limited by the parent environment".  Before this KEP the parent of a container is effectively either the node or the QoS group (since you can't set the pod explicitly).  After this KEP the parent is the pod (if specified) or the node or QoS group.

Defining it this way makes it clear that the following are obviously valid:
- set container limits/requests and no pod limit/request -> pod value is inferred to be sum(containers)
- set pod limit/request and no container limits/requests -> container value == pod value, no isolation within

What isn't clear is what happens if container requests or limits are > pod.

>  Each Linux cgroup behaves differently in cases where the lower-level requests exceed a higher point in the hierarchy.

We should be really careful NOT to tie our API to "whatever linux cgroups do".  Partly because we have non-Linux nodes, partly because we have non-cgroup implementations, and partly because linux cgroups are always evolving.

I think we should define (and enforce) this case for both request and limit.  Specifically, for both request and limit (and the answers may be different for each): 

* Do we allow a single container whose value exceeds the pod's value?  If so, what semantics are guaranteed and what are "implementation defined"?

* Do we allow the sum of containers' values to exceed the pod's value (even if each container does not)?  If so, what semantics are guaranteed and what are "implementation defined"?

Put another way - maybe we should start by saying these cases are not allowed, which gives us a clearer and easier road to GA.  THEN we can define these if and when the need arises.

If we don't define it, we will be stuck supporting whatever craziness happens to work on users's systems.

Reply:
> We should document this really clearly. We frequently say or imply that not setting a limit means "no limit". What it really > means is "limited by the parent environment". Before this KEP the parent of a container is effectively either the node or the QoS group (since you can't set the pod explicitly). After this KEP the parent is the pod (if specified) or the node or QoS group.

This is an extremely important point. But to make it more accurate, even before this KEP the limit is not only the parent environment, it's the entire path of parents from the pod all the way to the cgroup hierarchy root. Even before this KEP above the `kubepods` level in the hierarchy there are additional levels. This is why it's possible to [reserve resources for system daemons](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#kube-reserved). 

I completely agree that this should be well documented. Even before this KEP is approved or implemented.

> We should be really careful NOT to tie our API to "whatever linux cgroups do".

While I agree that the API must not be tied to a specific implementation, it's not going to be possible to define semantics that don't take the underlying implementation into consideration. When linux cgroups evolve, Kubernetes will have to evolve with them. And if the semantics we define don't agree with an alternate implementation, the semantics are going to have to change, or the alternate platform won't be supportable. 

Looking at the Windows `jobs` implementation, there are [similar semantics](https://docs.microsoft.com/en-us/windows/win32/procthread/nested-jobs) there as well for hierarchical resource controls.

Fortunately, the Linux kernel community is very strict that there are no behavior changes to published user-space interfaces. I'm guessing that Windows has a similar guarantee.

Can you link any other non cgroups/windows-jobs implementations that have different semantics?

> What isn't clear is what happens if container requests or limits are > pod.

The problem I see is that it's a well-accepted practice to allow admission webhooks to modify incoming resource requests. This means that the resources explicitly defined by the developer for a pod might not be enough to get a pod running. For example, consider a pod where the limits are exactly the sum of the container limits when the developer submits it to Kubernetes. Then an admission controller injects an additional sidecar and includes a request/limit in the sidecar definition. The developer has no way of making this work without trial and error to see how much slack should be left for this injection. And next time the admission controller version changes, some of the pods running on the cluster could mysteriously fail to run due to changes in the new version's choice of request/limit values. 

This _could_ be solved by updating all admission controllers to _also_ update pod level resources - but that would both require all the eco-system to update controllers or stop working, and override the developer's intent in a possibly incompatible way.

Semantically, I see setting a pod-level as defining a higher-level restriction on a pod. For example, in my use-case, where the nodes are over-provisioned and each pod potentially contains a different number of sidecars, the pod-level limit is a way of defining a resource budget for the entire pod. In order to pack `n` pods into a node without micro-managing the resource split between the containers, a pod-level limit is defined. Since the number of sidecars is dependent on choices an end-user makes, it would be very harmful for my use-case if a pod is rejected when `sum(container limits) > pod-level limit`. And if any admission controllers would change the predefined pod-level budget under the hood, then the system as a whole would become hard to size correctly.

As a user, I'd like to set my pod-level restriction/budget and have the system just "figure it out" to do the best it can. 

>  *   Do we allow a single container whose value exceeds the pod's value? If so, what semantics are guaranteed and what are "implementation defined"?
>  *  Do we allow the sum of containers' values to exceed the pod's value (even if each container does not)? If so, what semantics are guaranteed and what are "implementation defined"?

I think that the semantics are actually made very clear with the statement from the beginning of this thread:
> What it really means is "limited by the parent environment"

Not to mention that the resource limit is actually (even today) defined differently across resource type. For example, CPU CFS shares are defined relative to the parent environment. Which means (before this KEP as well) that a "1 CPU" limit on a container in a `Burstable` pod means different things based on the number of `Guaranteed` pods running on the node.

Reply:
This reply got long, sorry!

>> We should document this really clearly. We frequently say or imply that not setting a limit means "no limit". What it really means is "limited by the parent environment". Before this KEP the parent of a container is effectively either the node or the QoS group (since you can't set the pod explicitly). After this KEP the parent is the pod (if specified) or the node or QoS group.
>
> This is an extremely important point. But to make it more accurate, even before this KEP the limit is not only the parent environment, it's the entire path of parents from the pod all the way to the cgroup hierarchy root. Even before this KEP above the `kubepods` level in the hierarchy there are additional levels. This is why it's possible to [reserve resources for system daemons](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#kube-reserved). 

I think we have been careful to make sure those parent-level are not over-committed, haven't we?

> I completely agree that this should be well documented. Even before this KEP is approved or implemented.

Good.  We can always loosen spec, but never tighten it.

>> We should be really careful NOT to tie our API to "whatever linux cgroups do".
>
> While I agree that the API must not be tied to a specific implementation, it's not going to be possible to define semantics that don't take the underlying implementation into consideration. When linux cgroups evolve, Kubernetes will have to evolve with them. And if the semantics we define don't agree with an alternate implementation, the semantics are going to have to change, or the alternate platform won't be supportable. 

I understand what you are saying, but I want to disagree with the conclusion.  We should tightly spec the behaviors we want to allow and disallow behaviors that we can't tightly spec, until such time that we CAN spec them or prove that looser specs are safe.

In other words - I think we can define pretty precisely what it means for a pod to have a limit greater than the sum of the containers' requests.  We might need to leave a small escape clause for "system-defined" but maybe we don't.

Really concretely, I think we need to define these three example cases.

Case 1:
 - c1:  { requests[cpu]: 2 }
 - c2:  { requests[cpu]: 2 }
 - pod: { limits[cpu]: 4 }

I think we pretty clearly just need to specify the behavior we want. Both containers are guaranteed 2 cores but can burst to 4 cores with equal availability.  This could be implemented as inferring that c1 and c2 have a limit of 4.

Case 2:
 - c1:  { requests[cpu]: 2 }
 - c2:  { requests[cpu]: 2 }
 - pod: { limits[cpu]: 6 }

Again, I think it's pretty clear what this means.  Both containers are guaranteed 2 cores but can burst to 6 cores with equal availability.

Case 3:
 - c1:  { requests[cpu]: 1 }
 - c2:  { requests[cpu]: 2 }
 - pod: { limits[cpu]: 6 }

Now it gets interesting.  Imagine both containers are using as much CPU as they can.  Do they:
a) level out at 3/3
b) level out at 2.5/3.5 (extra CPU is evenly shared)
c) level out at 2/4 (extra CPU is proportionally shared)
d) we define this as "implementation defined"

I think all 3 could be achieved by setting the per-container shares.  What would users be least shocked by?

> Looking at the Windows `jobs` implementation, there are [similar semantics](https://docs.microsoft.com/en-us/windows/win32/procthread/nested-jobs) there as well for hierarchical resource controls.
>
> Fortunately, the Linux kernel community is very strict that there are no behavior changes to published user-space interfaces. I'm guessing that Windows has a similar guarantee.
>
> Can you link any other non cgroups/windows-jobs implementations that have different semantics?

How about non-cgroup runtimes?  Kata?  GVisor?  Fargate?

In short, I am saying not to define this in terms of cgroups.  Define the behaviors we think are reasonable and least-surprising.

>> What isn't clear is what happens if container requests or limits are > pod.
>
> The problem I see is that it's a well-accepted practice to allow admission webhooks to modify incoming resource requests. This means that the resources explicitly defined by the developer for a pod might not be enough to get a pod running. For example, consider a pod where the limits are exactly the sum of the container limits when the developer submits it to Kubernetes. Then an admission controller injects an additional sidecar and includes a request/limit in the sidecar definition. The developer has no way of making this work without trial and error to see how much slack should be left for this injection. And next time the admission controller version changes, some of the pods running on the cluster could mysteriously fail to run due to changes in the new version's choice of request/limit values. 
>
> This _could_ be solved by updating all admission controllers to _also_ update pod level resources - but that would both require all the eco-system to update controllers or stop working, and override the developer's intent in a possibly incompatible way.

Excellent points.  We can't break the ecosystem,  But we also don't want to modify user-inputs (usually).  But we also don't want to start breaking pods that used to work.

Forcing sidecar-injecting webhooks to update pod resources is a breaking change, ergo we *must* DTRT when a sidecar is injected without updating pod resources.

Changing the user's input is, at best open to interpretation.  If the user gave us:
 - c1:  { requests[cpu]: 1 }
 - c2:  { requests[cpu]: 2 }
 - pod: { limits[cpu]: 6 }

Did they mean "3 extra CPUs" or "200% of the sum"?  If I inject c3 which requests 1 CPU does that mean:
a) it fits the pod, no need to change anything
b) the pod limit should change to 7 (sum+3)
c) the pod limit should change to 8 (sum*2)

And even if we pick an interpretation, at what point of the lifecycle do we know which parts the user specified (c1 & c2 & pod, but not c3) so we can make auto-adjustments?  I don't think we have a place to do that?

Not changing the pod limit means that, if the user specified a pod limit, we start over-committing.  Best case the pod just fails in a spectacular explosion.  Worst case it experiences degraded performance and long-tail latency.  Not a great path either.

> Semantically, I see setting a pod-level as defining a higher-level restriction on a pod. For example, in my use-case, where the nodes are over-provisioned and each pod potentially contains a different number of sidecars, the pod-level limit is a way of defining a resource budget for the entire pod. In order to pack `n` pods into a node without micro-managing the resource split between the containers, a pod-level limit is defined. Since the number of sidecars is dependent on choices an end-user makes, it would be very harmful for my use-case if a pod is rejected when `sum(container limits) > pod-level limit`. And if any admission controllers would change the predefined pod-level budget under the hood, then the system as a whole would become hard to size correctly.

So if a cluster admin decides to add a new sidecar injector, user pods will start misbehaving, OOMing, or just slowing down.  That's a breaking change.  Worse - cluster admin didn't change how many sidecars, just upgraded the sidecar (e.g. let's pick on Istio).  If version 1 of the sidecar needed 1 core and version 2 needed 2 cores, I just subtracted a whole core from every pod's actually-usable resources!

> As a user, I'd like to set my pod-level restriction/budget and have the system just "figure it out" to do the best it can. 

I don't think we CAN figure it out.  All paths are ambiguous and/or breaking.

Let's try a different idea.  What if, instead of pod-resources being absolute numbers, they could be deltas?  E.g. "my pod's CPU limit is sum(container.request) + 2.0" ?  That doesn't do great if you don't specify pod values, and it would have to cross between requests and limits.

We should attack this from other angles - see if we can find something that works better (or at least convince ourselves we tried and failed to do better).

>>  *   Do we allow a single container whose value exceeds the pod's value? If so, what semantics are guaranteed and what are "implementation defined"?
>>  *  Do we allow the sum of containers' values to exceed the pod's value (even if each container does not)? If so, what semantics are guaranteed and what are "implementation defined"?
>
> I think that the semantics are actually made very clear with the statement from the beginning of this thread:
>> What it really means is "limited by the parent environment"
>
> Not to mention that the resource limit is actually (even today) defined differently across resource type. For example, CPU CFS shares are defined relative to the parent environment. Which means (before this KEP as well) that a "1 CPU" limit on a container in a `Burstable` pod means different things based on the number of `Guaranteed` pods running on the node.

Reply:
This reply is also long, sorry as well! 

> We should tightly spec the behaviors we want to allow and disallow behaviors that we can't tightly spec 

This is really hard. And if I'm not mistaken, the three cases you defined wouldn't actually work the way you outlined.

The issue is that Kubernetes uses the `request` part of the resources section as input to the scheduler. None of this information is passed down to the runtime. Only limits are actually enforced - requests are not. There is no concept of resource reservation.

Today you can approximate CPU and memory reservations in the way you described only in a very specific usage profile where the administrator can guarantee that all of these conditions are true:
1. There is zero resource over-provisioning on a node
1. All of the pods running on a node are in the Guaranteed QoS group

The `CPUManager` can also allow you to allocate CPU cores to a (today) specific container, actually reserving it for that container. You need to enable the static profile for this, and it only works for whole units of CPU cores - so for example you can't reserve 0.5 CPU cores to a container - it gets rounded up to the nearest full unit.

> Case 1:
>
>    c1: { requests[cpu]: 2 }
>    c2: { requests[cpu]: 2 }
>    pod: { limits[cpu]: 4 }

> I think we pretty clearly just need to specify the behavior we want. Both containers are guaranteed 2 cores but can burst > to 4 cores with equal availability. This could be implemented as inferring that c1 and c2 have a limit of 4.

But this wouldn't actually be the result in the most common case with the most common runtime. In Linux there is no real way to _reserve_ fractional amounts of CPU, or any amount of memory before it's actually allocated by a process. Only a way to _limit_ the amount of CPU or memory being consumed from exceeding a certain amount. The behavior we want may be to reserve 2 cores to each container, but the behavior that will actually be provided is that the entire pod will be limited to 4 CPUs, and both c1 and c2 can use any fraction of those 4 CPUs at any given time. If c1 is running a busy-loop, then c2 can be CPU starved and not receive any CPU shares at all.

Same thing goes for memory - it's possible to limit a container or a pod from allocating more than `n` bytes of memory, but until a process actually allocates that memory, the memory is not _reserved_. In this example (assume GiB of memory instead of cpu count), both c1 and c2 could allocate up to the sum of 4GiB, but there is no guarantee that 2GiB would be available to c1 or to c2. 

It doesn't get better if you switch runtimes. (I actually ran kata and gVisor to see what happens, but I don't have access to a container enabled windows machine, or to a working Fargate installation):

*. The kata container runtime already today allocates a pseudo pod-level limit to the containing VM (you can control this today with [annotations](https://github.com/kata-containers/documentation/blob/master/how-to/how-to-set-sandbox-config-kata.md) on the pod level), and then runs a [container manager]( https://github.com/kata-containers/documentation/blob/master/design/architecture.md ) based on runc inside that VM that managed the individual containers with cgroups. But there is no actual CPU reservation going on here on the system level. 

*. gVisor actually ignores the container level memory and cpu limits altogether. See https://github.com/google/gvisor/issues/190 for details. As with standard cgroups, any of the containers can starve any of the other containers in this pod of CPU and of memory. Even worse, when you run a pod with gVisor that _should_ be classified as `Guaranteed`, it's actually going to be run as `Burstable` and there will be no error message or warning anywhere.

If you didn't have pod level limits, but rather put the limits (as today) on the container level, then this case is impossible to model as a guaranteed pod. Even as a burstable pod it's not clear if it's possible. The requirement here is to limit each container to no more than 4 cpus, but allow any one of them to use 4 cpus as a maximum, while sharing the available cpus with the other containers in the pod. Without pod-level limits - I don't see how this can be expressed.

> Case 2:
>
>    c1: { requests[cpu]: 2 }
>    c2: { requests[cpu]: 2 }
>    pod: { limits[cpu]: 6 }
>
> Again, I think it's pretty clear what this means. Both containers are guaranteed 2 cores but can burst to 6 cores with equal availability.

Actually none of the containers are guaranteed any amount of cores. Either of the containers in the pod can starve the other with a busy loop. It's pretty much the same as the above example. Requests are used to enable Kubernetes to schedule the pod to a node, but once that's done, then the requests don't matter any more. AFAICT, in any runtime. 

This case is basically exactly the same as the first, due to it being impossible to _reserve_ CPU or memory without actually allocating the memory or actively using the cpu.

> Case 3:
> 
>    c1: { requests[cpu]: 1 }
>    c2: { requests[cpu]: 2 }
>    pod: { limits[cpu]: 6 }

> Now it gets interesting. Imagine both containers are using as much CPU as they can. Do they:
> a) level out at 3/3
> b) level out at 2.5/3.5 (extra CPU is evenly shared)
> c) level out at 2/4 (extra CPU is proportionally shared)
> d) we define this as "implementation defined"

Since it's impossible to reserve CPUs for a container, the containers would not level out anywhere in this example. So a,b, and c are all impossible. Based on how much CPU each container is requesting, that container would get all the CPU it needs. The entire pod would be _limited_ to 6 virtual cores since a limit is possible, a reservation isn't. So at no point in time will the sum of CPU used by all of these containers be greater than 6.

This means that the situation today (with container-level limits) is already "implementation defined", and the "official" documentation today only really documents what happens when you use the standard cgroups implementation. 

> I think all 3 could be achieved by setting the per-container shares. What would users be least shocked by?

I'm really not sure that's true. Maybe the first option is possible by limiting each container to 3 cores, in which case the pod-level resource setting is really moot.

But the other options are all impossible on Linux, since CPU reservation is not supported on Linux. I doubt even using a different runtime would help here, unless that different runtime was some sort of RTOS, and you updated the CRI interface to be aware of the relationship between two different containers.

> We should attack this from other angles - see if we can find something that works better (or at least convince ourselves we tried and failed to do better).

My intuitive understanding here is that we go with expanding on the pod definition. Kubernetes already defines the pod to be a scheduling unit. Let's extend that to resource requests. The rules become:

1. If no pod-level limits are defined - behave as before this KEP.
1. If pod level limits are defined:
    1. Pod level _requests_ are used as the input for the scheduler for placing the entire pod on a node. 
    1. Pod level _limits_ are considered a upper limit for the resources of all of the containers in the pod. This expresses an approved budget for the entire pod. This also works for Kata and for gVisor even today - e.g. Kata allows defining this using pod-level annotations.
    1. Container level _requests_ are either ignored (e.g. for the CPU resource) or used as a hint (e.g. for setting the soft_limit for the memory resource to improve the Linux OOM handler's behavior).
    1. Container level _limits_ are set on the container level (specific runtime permitting). 
1. The documentation is updated to explain the meaning of setting limits on each level of the hierarchy. The requirement here is for a hierarchical resource limit definition, and the implementation details are left to each runtime to manage. 

The axiom is that each limit in a specific level in the resource hierarchy is always defined in relation to the level above it. This is true today, and allowing developers to specify pod-level resources doesn't change this.

By defining pod-level resources like this, it becomes easier to understand what is happening. 

For CPU: if `pod-cpu-limit < sum(containers-cpu-limit)`, then the pod-level CPU limit is higher in the hierarchy and therefore is considered the real limit. The container level limits will define the proportional split of CPU quota between containers since the pod is the level above the containers. Before this KEP the exact same thing happened, except that it was in relation to the QoS group, so this isn't really an API change. 

For memory: If `pod-memory-limit < sum(containers-memory-limit)`, then there is going to be an OOM situation here due to the way memory allocation works. But this is exactly the same as the situation today, except that today the granularity level without pod-level resources was QoS groups and not pods.

Reply:
First - awesome discussion, thanks!

> The issue is that Kubernetes uses the `request` part of the resources section as input to the scheduler. None of this information is passed down to the runtime. Only limits are actually enforced - requests are not. There is no concept of resource reservation.

What about CPU shares?  Thsat reflects requests.  It might be true for memory - for now - but it is not universally true.  You can trivially test this and then poke into cgroupfs.

> Today you can approximate CPU and memory reservations in the way you described only in a very specific usage profile where the administrator can guarantee that all of these conditions are true:
> 1. There is zero resource over-provisioning on a node
> 1. All of the pods running on a node are in the Guaranteed QoS group

As long as we don't overcommit shares at any level, we know that even if EVERY pod on the machine burns as much CPU as it can, it will settle near the `request`.  In a pod whose request is greater than the sum of the containers, that "extra" CPU is supposed to be reserved for the member containers.  

Our job, I think, is to spec the behavior we want and then figure out how to deliver that.  Which may include fudging the member containers' cpu.shares.

> The `CPUManager` can also allow you to allocate CPU cores to a (today) specific container, actually reserving it for that container. You need to enable the static profile for this, and it only works for whole units of CPU cores - so for example you can't reserve 0.5 CPU cores to a container - it gets rounded up to the nearest full unit.

That's fine, I think.  If you `request` a non-integer number of cores, you get some N reserved and some shared.  Again, let's define semantics, not implementation.

>> Case 1:
>>
>>    c1: { requests[cpu]: 2 }
>>    c2: { requests[cpu]: 2 }
>>    pod: { limits[cpu]: 4 }
>
>> I think we pretty clearly just need to specify the behavior we want. Both containers are guaranteed 2 cores but can burst > to 4 cores with equal availability. This could be implemented as inferring that c1 and c2 have a limit of 4.
>
> But this wouldn't actually be the result in the most common case with the most common runtime. In Linux there is no real way to _reserve_ fractional amounts of CPU, or any amount of memory before it's actually allocated by a process. Only a way to _limit_ the amount of CPU or memory being consumed from exceeding a certain amount. The behavior we want may be to reserve 2 cores to each container, but the behavior that will actually be provided is that the entire pod will be limited to 4 CPUs, and both c1 and c2 can use any fraction of those 4 CPUs at any given time. If c1 is running a busy-loop, then c2 can be CPU starved and not receive any CPU shares at all.

That's not true, I think.  Assuming 1024 shares per CPU. c1 and c2 should have 2048 shares each, which means that in case of a busy-loop they should approach equal CPU usage.

If we can't guarantee that much, we are really missing the point.

> Same thing goes for memory - it's possible to limit a container or a pod from allocating more than `n` bytes of memory, but until a process actually allocates that memory, the memory is not _reserved_. In this example (assume GiB of memory instead of cpu count), both c1 and c2 could allocate up to the sum of 4GiB, but there is no guarantee that 2GiB would be available to c1 or to c2. 

ACK that memory is different.  Hence we should describe semantics we want, that make the most sense for users, and try to implement them as closely as we can.  We may need to split our spec by resource or by compressible/incompressible.

> *. The kata container runtime already today allocates a pseudo pod-level limit to the containing VM (you can control this today with [annotations](https://github.com/kata-containers/documentation/blob/master/how-to/how-to-set-sandbox-config-kata.md) on the pod level), and then runs a [container manager]( https://github.com/kata-containers/documentation/blob/master/design/architecture.md ) based on runc inside that VM that managed the individual containers with cgroups. But there is no actual CPU reservation going on here on the system level. 

The CPU "reservation" is requests.

> *. gVisor actually ignores the container level memory and cpu limits altogether. See https://github.com/google/gvisor/issues/190 for details. As with standard cgroups, any of the containers can starve any of the other containers in this pod of CPU and of memory.  

Bug!

> If you didn't have pod level limits, but rather put the limits (as today) on the container level, then this case is impossible to model as a guaranteed pod. Even as a burstable pod it's not clear if it's possible. The requirement here is to limit each container to no more than 4 cpus, but allow any one of them to use 4 cpus as a maximum, while sharing the available cpus with the other containers in the pod. Without pod-level limits - I don't see how this can be expressed.

shares = 2048; quota = 4 (whatever math there) - this gives equal weight to the 2 containers.  At least, that's how I understand it is supposed to work.

>> Case 2:
>>
>>    c1: { requests[cpu]: 2 }
>>    c2: { requests[cpu]: 2 }
>>    pod: { limits[cpu]: 6 }
>>
>> Again, I think it's pretty clear what this means. Both containers are guaranteed 2 cores but can burst to 6 cores with equal availability.
>
> Actually none of the containers are guaranteed any amount of cores. Either of the containers in the pod can starve the other with a busy loop. It's pretty much the same as the above example. Requests are used to enable Kubernetes to schedule the pod to a node, but once that's done, then the requests don't matter any more. AFAICT, in any runtime. 

What you describe is totally NOT how I understand and recall the kernel working here.  shares should "guarantee" that, on average, other containers get time.  If everyone runs full-tilt, usage should be ~= shares-proprtional.

You can see cpu.shares being set at both the pod and container levels just by inspecting a running pod via cgroupfs

>> Case 3:
>> 
>>    c1: { requests[cpu]: 1 }
>>    c2: { requests[cpu]: 2 }
>>    pod: { limits[cpu]: 6 }
>
>> Now it gets interesting. Imagine both containers are using as much CPU as they can. Do they:
>> a) level out at 3/3
>> b) level out at 2.5/3.5 (extra CPU is evenly shared)
>> c) level out at 2/4 (extra CPU is proportionally shared)
>> d) we define this as "implementation defined"
>
> Since it's impossible to reserve CPUs for a container, the containers would not level out anywhere in this example. So a,b, and c are all impossible. Based on how much CPU each container is requesting, that container would get all the CPU it needs. The entire pod would be _limited_ to 6 virtual cores since a limit is possible, a reservation isn't. So at no point in time will the sum of CPU used by all of these containers be greater than 6.

After the above - if I am still missing some point, I apologize.  This topic was one that Borg wrestled with, so I am 98% sure it's expressible in cgroups.  Specifically, assuming the pod has a shares of 6*1024:

a) set shares to 3*1024, 3*1024
b) set shares to 2.5*1024, 3.5*1024
c) set shares to 2048, 4096

We just need to define what we think is correct behavior.

>> I think all 3 could be achieved by setting the per-container shares. What would users be least shocked by?
>
> I'm really not sure that's true. Maybe the first option is possible by limiting each container to 3 cores, in which case the pod-level resource setting is really moot.

I was maybe overly tricky here - I set pod limits but not requests, knowing that (as per API docs) "If Request is omitted for a container, it defaults to Limits if that is explicitly specified"

So if I set the pod request to 6 cores and limit to 6 cores, it *is* guaranteed (or should be) and the scheduler (unless I deeeeeply misunderstand it) will consider the pod for 6*1024 shares.  When the scheduler selects the pod cgroup, it starts over and looks at the shares of all the members within it.  *we* define the balance between c1 and c2 via shares.

This is the part where you tell me my memory is wrong (and all the googling I just did is wrong too :)  If so, I humbly beg forgiveness, and we'll have to maybe rewind.

>> We should attack this from other angles - see if we can find something that works better (or at least convince ourselves we tried and failed to do better).
>
> My intuitive understanding here is that we go with expanding on the pod definition. Kubernetes already defines the pod to be a scheduling unit. Let's extend that to resource requests. The rules become:

Given the very fundamental difference in understanding above, I can't really speak to this section yet, except to say it still sounds like a breaking change and I think we can do better.

> 1. If no pod-level limits are defined - behave as before this KEP.
> 1. If pod level limits are defined:
>    1. Pod level _requests_ are used as the input for the scheduler for placing the entire pod on a node. 
>    1. Pod level _limits_ are considered a upper limit for the resources of all of the containers in the pod. This expresses an approved budget for the entire pod. This also works for Kata and for gVisor even today - e.g. Kata allows defining this using pod-level annotations.
>    1. Container level _requests_ are either ignored (e.g. for the CPU resource) or used as a hint (e.g. for setting the soft_limit for the memory resource to improve the Linux OOM handler's behavior).
>    1. Container level _limits_ are set on the container level (specific runtime permitting). 
> 1. The documentation is updated to explain the meaning of setting limits on each level of the hierarchy. The requirement here is for a hierarchical resource limit definition, and the implementation details are left to each runtime to manage. 
>
> The axiom is that each limit in a specific level in the resource hierarchy is always defined in relation to the level above it. This is true today, and allowing developers to specify pod-level resources doesn't change this.
>
> By defining pod-level resources like this, it becomes easier to understand what is happening. 
>
> For CPU: if `pod-cpu-limit < sum(containers-cpu-limit)`, then the pod-level CPU limit is higher in the hierarchy and therefore is considered the real limit. The container level limits will define the proportional split of CPU quota between containers since the pod is the level above the containers. Before this KEP the exact same thing happened, except that it was in relation to the QoS group, so this isn't really an API change. 
>
> For memory: If `pod-memory-limit < sum(containers-memory-limit)`, then there is going to be an OOM situation here due to the way memory allocation works. But this is exactly the same as the situation today, except that today the granularity level without pod-level resources was QoS groups and not pods.

Reply:
Given a 2 core machine, I ran a quick test:

```
# Test playground
mkdir /sys/fs/cgroup/cpu/test
echo $$ > /sys/fs/cgroup/cpu/test/tasks

# Fake a pod with 2 containers
mkdir /sys/fs/cgroup/cpu/test/pod
mkdir /sys/fs/cgroup/cpu/test/pod/c1
mkdir /sys/fs/cgroup/cpu/test/pod/c2

# Allocate 0.6 for the pod and 0.3 for each container
echo $((6*1024/10)) > /sys/fs/cgroup/cpu/test/pod/cpu.shares
echo $((3*1024/10)) > /sys/fs/cgroup/cpu/test/pod/c1/cpu.shares
echo $((3*1024/10)) > /sys/fs/cgroup/cpu/test/pod/c2/cpu.shares

# Run a hot loop in c1
cp /bin/yes /var/run/docker/c1
/var/run/docker/c1 >/dev/null &
echo $! > /sys/fs/cgroup/cpu/test/pod/c1/tasks

# Run a hot loop in c2
cp /bin/yes /var/run/docker/c2
/var/run/docker/c2 >/dev/null &
echo $! > /sys/fs/cgroup/cpu/test/pod/c2/tasks

# Run something else to burn CPU, using the other 1.4 cores
mkdir /sys/fs/cgroup/cpu/test/other
echo $((14*1024/10)) > /sys/fs/cgroup/cpu/test/other/cpu.shares
yes >/dev/null &
echo $! > /sys/fs/cgroup/cpu/test/other/tasks
yes >/dev/null &
echo $! > /sys/fs/cgroup/cpu/test/other/tasks

# Now look at top - c1 and c2 settle around 30%.

# Change it
echo $((2*1024/10)) > /sys/fs/cgroup/cpu/test/pod/c1/cpu.shares
echo $((4*1024/10)) > /sys/fs/cgroup/cpu/test/pod/c2/cpu.shares

# Now look at top - c1 and c2 settle around 20% and 40%
```

I think this proves it is possible?

Reply:
What this shows is that CPU will be scheduled to each task fairly and no CPU starvation occurs. 

It's not _exactly_ the same as CPU reservation, which is what the `cpuset` cgroup provides. The main difference being that when a CPU is reserved to a set of tasks, it is never allocated to a different set of tasks. This is critical for some workloads because of  L1 and L2 caches on the CPU itself. 

But I'm nitpicking. I agree that for most workloads this is a close enough approximation of CPU reservation.

Reply:
Looking back at the case 3 you defined:
>Case 3:
>
>    c1: { requests[cpu]: 1 }
>    c2: { requests[cpu]: 2 }
>    pod: { limits[cpu]: 6 }
>
> Now it gets interesting. Imagine both containers are using as much CPU as they can. Do they:
> a) level out at 3/3
> b) level out at 2.5/3.5 (extra CPU is evenly shared)
> c) level out at 2/4 (extra CPU is proportionally shared)
> d) we define this as "implementation defined"

Since the shares in each level is are distributed relative to the parent level in the hierarchy, wouldn't that mean that the default behavior here would be (c) ?  The entire pod is limited to 6 CPUs, which are more-or-less reserved for it. Each container in the pod gets a proportional share relative to the pod maximum.  And adding an additional container in the pod (e.g. because of a sidecar injection) is then well defined.

The difficulty becomes that while the pod level represent actual CPUs, the shares set at the container level don't necessarily represent CPU units. For example, the sum of the container limits might (due to unexpected sidecar injection) be greater than the limit defined in the pod. This is still a valid configuration, simply because the container level is set proportionally to the pod level. Ref: https://www.kernel.org/doc/Documentation/scheduler/sched-design-CFS.rst 

I think this is still OK, especially since the entire pod is what Kubernetes was asked to run. Treating the pod level as the "resource budget" and the internals as a proportional split would surprise me the least in this case.

Reply:
Sorry I was AFK - I have more time the next few weeks - let's get to
conclusions :)  If you want to talk in real-time, we can do that.  I am US-West
timezone.

> Looking back at the case 3 you defined:
>
>>Case 3:
>>
>>    c1: { requests[cpu]: 1 }
>>    c2: { requests[cpu]: 2 }
>>    pod: { limits[cpu]: 6 }
>>
>> Now it gets interesting. Imagine both containers are using as much CPU as they can. Do they:
>> a) level out at 3/3
>> b) level out at 2.5/3.5 (extra CPU is evenly shared)
>> c) level out at 2/4 (extra CPU is proportionally shared)
>> d) we define this as "implementation defined"
>
> Since the shares in each level are distributed relative to the parent level in the hierarchy, wouldn't that mean that the default behavior here would be (c) ?  The entire pod is limited to 6 CPUs, which are more-or-less reserved for it. Each container in the pod gets a proportional share relative to the pod maximum.  And adding an additional container in the pod (e.g. because of a sidecar injection) is then well defined.

I think you are correct that, if we specify nothing else, that is what will
happen.  HOWEVER... we can *choose* to define it how we want.  We can
explicitly apportion the "extra" so that it comes out however we want.  So I
think we should ask what behavior makes the most sense, rather than what
the kernel does by default.

Of course, CPU and memory are different, and there's not (currently) any way to
squeeze one container to free up memory for the other.  But you can imagine a
kernel API for that, one day - so I think the question remains -- what behavior
makes the most sense?

> The difficulty becomes that while the pod level represent actual CPUs, the shares set at the container level don't necessarily represent CPU units. For example, the sum of the container limits might (due to unexpected sidecar injection) be greater than the limit defined in the pod. This is still a valid configuration, simply because the container level is set proportionally to the pod level. Ref: https://www.kernel.org/doc/Documentation/scheduler/sched-design-CFS.rst

Yes, it is valid at the kernel level.  The question, again, is whether we want
to define Kubernetes API and semantics in terms of that?

> I think this is still OK, especially since the entire pod is what Kubernetes was asked to run. Treating the pod level as the "resource budget" and the internals as a proportional split would surprise me the least in this case.

I see two paths, I think.

1) As you describe, the outer pod limit is an absolute value and the alloctions
inside are relative to each other and bounded by the pod.

2) The outer pod limit is NOT an absolute value in the API, but an "extra" on
top of the sum of member containers.

The problem I have with (1) is what I described above.  It's a long thread so I
will restate it :)

We have some rules we try to live by:
* We can't make users suddenly start failing (breaking changes)
* We can't break the ecosystem
* We don't (usually) modify user-input

Assume a user has a multi-container pod that specifies a memory limit, and that
their app is well-tuned to stay within that limit.  Someone decides to inject a
sidecar.  The pod has a fixed limit.  The sidecar eats into that.  The user app
suddenly starts failing.  We could add the sidecar's resources to the pod, but
then we would be modifying user input.

The problem is that the API is ambiguous.  Does the user intend pod resources
to be a fixed value or did they intend for it to be "in addition to" the sum of
the containers?

I am not convinced this is an API that can work.

What if, instead of `pod.spec.Resources` as real values, we had
`pod.spec.ExtraResources` as offsets?

e.g. Instead of:

```
c1:  { resources: { requests[cpu]: 1 }}
c2:  { resources: { requests[cpu]: 2 }}
pod: { resources: { limits[cpu]: 6 }}
```

We had:

```
c1:  { resources: { requests[cpu]: 1 }}
c2:  { resources: { requests[cpu]: 2 }}
pod: { extraResources[cpu]: 3 }
```

Now the API is unambiguous.  A sidecar that gets injected does not start
crashing apps.  We don't modify the user input.

I ACK that this is a significant change from the current proposal, but I am not
convinced the proposal works.  I am very open to other models here, this is
just one idea.

To play it out further, let's think thru use cases

Note: the question about proportional vs even access to extra is orthoginal,
assume even access here.

## case 1: all containers specify no limits or requests

input:
```
c1:  request: "", limit: ""
c2:  request: "", limit: ""
pod: extraResources: 4
```

effective result:
```
c1:  request: 2, limit: inf
c2:  request: 2, limit: inf
pod: request: 4, limit: inf
```

## case 2: all containers specify requests, no limits

input:
```
c1:  request: 1, limit: ""
c2:  request: 1, limit: ""
pod: extraResources: 2
```

effective result:
```
c1:  request: 2, limit: inf
c2:  request: 2, limit: inf
pod: request: 4, limit: inf
```

## case 3: all containers specify limits, no requests

input:
```
c1:  request: "", limit: 1
c2:  request: "", limit: 1
pod: extraResources: 2
```

effective result:
```
c1:  request: 2, limit: 3
c2:  request: 2, limit: 3
pod: request: 4, limit: 4
```

## case 4: all containers specify limits == requests

input:
```
c1:  request: 1, limit: 1
c2:  request: 1, limit: 1
pod: extraResources: 2
```

effective result:
```
c1:  request: 2, limit: 3
c2:  request: 2, limit: 3
pod: request: 4, limit: 4
```

## case 5: all containers specify limits > requests

input:
```
c1:  request: 1, limit: 2
c2:  request: 1, limit: 2
pod: extraResources: 2
```

effective result:
```
c1:  request: 2, limit: 4
c2:  request: 2, limit: 4
pod: request: 4, limit: 6
```

## case 6: some containers specify limits > requests

input:
```
c1:  request: 1, limit: 2
c2:  request: "", limit: ""
pod: extraResources: 2
```

effective result:
```
c1:  request: 2, limit: 4
c2:  request: 1, limit: 2
pod: request: 3, limit: 4
```

One thing to not here is that it isn't possible in this simplified form to take
2 burtsable containers and make the pod guaranteed.  To do that we'd need to
make extra have both requests and limits.  Before I go figure that out, let's
discuss.

Reply:
Sorry for the long pause - I had some urgent work-related tasks.

> I see two paths, I think.
>
>  1.  As you describe, the outer pod limit is an absolute value and the allocations
>       inside are relative to each other and bounded by the pod.
> 
>  2. The outer pod limit is NOT an absolute value in the API, but an "extra" on
>       top of the sum of member containers.

The thing is, the reason I'm pursuing this concept is exactly to provide the first path. In our use-case, we are intentionally over-provisioning customer workloads on a cluster. This makes sense to us because our workload profile is that for over 75% of the time, any given pod will be idle, and resources are only required for a short period of time every so often. Our product provides an IDE in a pod, and developer think times are a lot more significant than the build operations that are occasionally run in the IDE.

Our pods have a lot of containers - depending on the specific use-case, it's between 7 and 14 containers per pod. We add containers to the pod to add additional tools to the environment and extend the IDE, based on the developer's needs. We initially tried to tailor each specific pod's limits and requests individually, but ran into big problems. Since we need to be able to deploy multiple pods to each host, we need hard guarantees that each pod will have a predictable resource limit, but if we want to enable different containers in the same pod to fluctuate with their resource consumption, we can't balance the various requirements on a container level.

For our use case, option 2 above is not tenable. It would absolutely not work for us to specify "extra" resources on the pod level. What we need is exactly the first option - we want to be able to **not** specify limits on any of the individual containers, and limit the entire pod on the pod level. Being able to do this will allow the resources to "flow" between containers in a much more dynamic way, while still allowing us to deterministically plan the capacity of our nodes, and our overprovisioning expectations for each node.

I can definitely see option 2 as an interesting additional option - I don't think the two options are mutually exclusive - but for our use-case it's definitely not a valid option.

> We have some rules we try to live by:
>
>    We can't make users suddenly start failing (breaking changes)
>    We can't break the ecosystem
>    We don't (usually) modify user-input

Agree with all of these rules. 

If option 1 is chosen - then this is a backwards compatible change. Users will not face any breaking change if they do nothing. If they did nothing - then that means they also didn't add a pod-level restriction. This means that the current behavior still applies - no changes. 

The ecosystem would not be broken either in this case. This is an opt-in functionality - and like any opt-in functionality if a developer opts in, then some tuning work is to be expected . For those cases where a developer has opted in to this functionality, the developer should be able to limit the pod based on his workload, and therefore will not be surprised.

The proposal currently doesn't explicitly modify user-input in any way.
 
> Assume a user has a multi-container pod that specifies a memory limit, and that
> their app is well-tuned to stay within that limit. Someone decides to inject a
> sidecar. The pod has a fixed limit. The sidecar eats into that. The user app
> suddenly starts failing. 

For our use-case, adding an additional sidecar that takes the pod to some extreme and breaks the pod would actually be the better solution. Product-wise, it's incentive for selling the user larger slices of resources. The fact that a pod has a fixed limit is actually a good feature - it means we can deterministically size our nodes correctly. We expect to over-provision our nodes, since the workload is idle 75% of the time. We treat resources the same way a bank lends more money than it actually has, on the assumption that not everybody will want their money at the same exact point in time.

> We could add the sidecar's resources to the pod, but then we would be modifying user input.

For our use-case modifying the use-input would be detrimental. We would very much prefer the pod to fail - and to fail consistently. The user (in our use-case) can react by buying a larger slice of resources, which would result in a larger pod-level resource budget.

> The problem is that the API is ambiguous. Does the user intend pod resources
> to be a fixed value or did they intend for it to be "in addition to" the sum of
> the containers?

At least in my mind, the API is not at all ambiguous. The pod (which is the Kubernetes lowest-level schedulable primitive) is assigned a fixed budget of resources. The intent of the developer is absolutely to define a fixed limit to the pod resource usage on the pod level. It is never intended to be "in addition to" the sum of the containers. 


I don't doubt that other products have different assumptions on their resource consumption. For systems that do not over-provision their nodes I can definitely see the value of your proposed option 2. And I would definitely move the definition of that option to from `pod.spec.Resources` to `pod.spec.ExtraResources` just so the two options can co-exist.

Reply:
> Sorry for the long pause - I had some urgent work-related tasks.

No problem!

> > I see two paths, I think.
> >
> >  1.  As you describe, the outer pod limit is an absolute value and the allocations
> >       inside are relative to each other and bounded by the pod.
> >
> >  2. The outer pod limit is NOT an absolute value in the API, but an "extra" on
> >       top of the sum of member containers.
>
> The thing is, the reason I'm pursuing this concept is exactly to provide the first path. In our use-case, we are intentionally over-provisioning customer workloads on a cluster. This makes sense to us because our workload profile is that for over 75% of the time, any given pod will be idle, and resources are only required for a short period of time every so often. Our product provides an IDE in a pod, and developer think times are a lot more significant than the build operations that are occasionally run in the IDE.

To be pedantic - you are over-provisioning (giving too much resources) or
over-committing (pretending you have more resources than you actually have)?  I
assume over-committing..

> Our pods have a lot of containers - depending on the specific use-case, it's between 7 and 14 containers per pod. We add containers to the pod to add additional tools to the environment and extend the IDE, based on the developer's needs. We initially tried to tailor each specific pod's limits and requests individually, but ran into big problems. Since we need to be able to deploy multiple pods to each host, we need hard guarantees that each pod will have a predictable resource limit, but if we want to enable different containers in the same pod to fluctuate with their resource consumption, we can't balance the various requirements on a container level.

And if you don't give it enough, it's OK to just crash loop?

> For our use case, option 2 above is not tenable. It would absolutely not work for us to specify "extra" resources on the pod level. What we need is exactly the first option - we want to be able to **not** specify limits on any of the individual containers, and limit the entire pod on the pod level. Being able to do this will allow the resources to "flow" between containers in a much more dynamic way, while still allowing us to deterministically plan the capacity of our nodes, and our overprovisioning expectations for each node.

Can't you model this as a very low request + "extra"?

E.g. I have 16 cores to use.  I want to run no more than 4 pods per core, so  I
set the per-container CPU request to 0 and the per-pod adder to 0.25.  If,
for some reason, EVERY pod started the compiler at the same time, they would
all get about 0.25 cores each, but any one of them *could* use more if is idle.


I'm sorry if this rehashes stuff you have already thought through.  If I am
misunderstanding, I will try to catch up fast.  I will be happy to be convinced
I am making a big deal of nothing.

> I can definitely see option 2 as an interesting additional option - I don't think the two options are mutually exclusive - but for our use-case it's definitely not a valid option.

Last resort - I'd prefer NOT to have 2 different APIs here, if we can avoid
it.

Thinking about how we can represent the intent with the least need for anyone
to understand the implementation or to think about subtle semantics.

If we could represent the use-case you described above as:
- N containers, each with no resources specified
- the pod, with your fixed-size request and limit "adders"

Doesn't that satisfy?  This makes me realize the semantics around "not
specified" and "unlimited" are a bit unclear.

> The ecosystem would not be broken either in this case. This is an opt-in functionality - and like any opt-in functionality if a developer opts in, then some tuning work is to be expected . For those cases where a developer has opted in to this functionality, the developer should be able to limit the pod based on his workload, and therefore will not be surprised.

The problem I am thinking through is when the developer is not in full control
- injected sidecars are defined by someone else but impact the user.

> > Assume a user has a multi-container pod that specifies a memory limit, and that
> > their app is well-tuned to stay within that limit. Someone decides to inject a
> > sidecar. The pod has a fixed limit. The sidecar eats into that. The user app
> > suddenly starts failing.
>
> For our use-case, adding an additional sidecar that takes the pod to some extreme and breaks the pod would actually be the better solution. Product-wise, it's incentive for selling the user larger slices of resources. The fact that a pod has a fixed limit is actually a good feature - it means we can deterministically size our nodes correctly. We expect to over-provision our nodes, since the workload is idle 75% of the time. We treat resources the same way a bank lends more money than it actually has, on the assumption that not everybody will want their money at the same exact point in time.

Again, that can be modelled as "zer per container plus X per pod" right?

> > We could add the sidecar's resources to the pod, but then we would be modifying user input.
>
> For our use-case modifying the use-input would be detrimental. We would very much prefer the pod to fail - and to fail consistently. The user (in our use-case) can react by buying a larger slice of resources, which would result in a larger pod-level resource budget.

I guess the decision of boosting the pod resources or not comes down to the
actor doing injection and how they are configured).  You would configure them
NOT to boost pod resources, someone else might choose differently.

> I don't doubt that other products have different assumptions on their resource consumption. For systems that do not over-provision their nodes I can definitely see the value of your proposed option 2. And I would definitely move the definition of that option to from `pod.spec.Resources` to `pod.spec.ExtraResources` just so the two options can co-exist.

Let me approach from a different angle, see if I can walk you into my thinking
(or walk myself out).

From a _purely_ API point of view, do you think it makes sense for a pod's
cpu request to be less than the sum of the constituent containers' cpu
requests?  Yes, I know we CAN implement it and we CAN define it.  But from a
"principle of least surprise" perspective, is it the best possible API?

I'm going to argue "no".  Given that we must handle that case by either:
a) failing such an API operation
b) treating such a pod's request to be the max()
c) preventing this from being expressible

On top of that, as above, every sidecar injector must either modify the pod's
request or not.

Case (a), sidecar does not modify the pod request: the pod can fail to create
(which is an ugly failure mode when you use a Deployment).

Case (a), sidecar does modify the pod request: the pod will only fail to create
in case of quota.

Case (b), I don't even want to entertain.  We don't modify user intent like
that.

Case (c).  The best way I can think of to avoid this expression is to make the
pod resource an adder.  This avoids the failure mode entirely, but still allows
us to express everything we need to express, I think.  Sidecar injectors don't
have to touch pod resources.

That said, case (a) is not the end of the world.  It just has an unfortunate
failure mode that I feel like we could avoid.

So it comes down, in my opinion, do a question of obviousness and failure modes
on the API, more than ambiguous semantics.

Thoughts?  If you want to jump on slack or a hangout this week or next, I am
open.